internal/plugin/connectors/tcp/mssql/connector.go

```diff
@@ -97,12 +99,109 @@ type SingleUseConnector struct {
 	backendConn net.Conn
 	clientConn  net.Conn
 	logger      log.Logger
+	// Note: We're following standard ctor naming practices with this field.
+	newMSSQLConnector types.MSSQLConnectorCtor
+	readPrelogin      types.ReadPreloginFunc
+	writePrelogin     types.WritePreloginFunc
+	readLogin         types.ReadLoginFunc
+	newTdsBuffer      types.TdsBufferCtor
+}
+
+// NewSingleUseConnector creates a new production SingleUseConnector
+func NewSingleUseConnector(logger log.Logger) *SingleUseConnector {
+	return NewSingleUseConnectorWithOptions(
+		logger,
+		NewMSSQLConnector,
+		ReadPreloginWithPacketType,
+		WritePreloginWithPacketType,
+		ReadLogin,
+		NewTdsBuffer,
+	)
+}
+
+// NewMSSQLConnector is the production implementation of MSSQLConnectorCtor,
+// used for creating mssql.Connector instances.  We need to wrap the raw
+// constructor provided by mssql (ie, mssql.NewConnector) in this function so
+// that it returns an interface, which enables us to mock it in unit tests.
+func NewMSSQLConnector(dsn string) (types.MSSQLConnector, error) {
+	c, err := mssql.NewConnector(dsn)
+	fn := func(ctx context.Context) (types.NetConner, error) {
+		driverConn, err := c.Connect(ctx)
+		if err != nil {
+			return nil, err
+		}
+		// This can never fail unless mssql package changes: panicking is fine
+		mssqlConn := driverConn.(types.NetConner)
+		return mssqlConn, nil
+	}
+	return types.MSSQLConnectorFunc(fn), err
+}
+
+// ReadPreloginWithPacketType is the production version of our readPrelogin
+// dependency, which delegates to the actual 3rd party driver.
+func ReadPreloginWithPacketType(
+	rawTdsBuffer interface{},
+	rawPktType interface{},
+) (map[uint8][]byte, error) {
+	// This can never fail unless mssql package changes: panicking is fine
+	tdsBuffer := rawTdsBuffer.(*mssql.TdsBuffer)
+	pktTypeInt := rawPktType.(int)
+	pktType := mssql.PacketType(pktTypeInt)
+	return mssql.ReadPreloginWithPacketType(tdsBuffer, pktType)
+}
+
+// WritePreloginWithPacketType is the production version of our writePrelogin
+// dependency, which delegates to the actual 3rd party driver.
+func WritePreloginWithPacketType(
+	rawTdsBuffer interface{},
+	fields map[uint8][]byte,
+	rawPktType interface{},
+) error {
+	// This can never fail unless mssql package changes: panicking is fine
+	tdsBuffer := rawTdsBuffer.(*mssql.TdsBuffer)
```

Comment:
1. I'm curious why rawTdsBuffer and rawPktType are not their expected types and instead of type `interface{}` ? 
2. What was the motivation for making `PacketType` public ? I ask mostly because that's a change that moves the fork further from the original

Reply:
> I'm curious why rawTdsBuffer and rawPktType are not their expected types and instead of type interface{}

Good question.  Short answer: Because then our doubles would have a dep on the 3rd party driver.

> What was the motivation for making PacketType public

Because for our production impl we have to do this:

```
return mssql.ReadPreloginWithPacketType(tdsBuffer, pktType)
```
and `mssql.ReadPreloginWithPacketType` requires an argument of that type.  We can't cast to it without having the type available here.  

If you seen alternative I'm all ears....